### PR TITLE
Swerve

### PR DESCRIPTION
**Pull request by NoahSimon8**
# SwerveSubsystem 
 - Notible change includes the AprilTagGetters file and switching from SwerveDriveOdometry to SwerveDrivePoseEstimator. It should now update its pose when spotting april tags

## Button Bindings
 - Pilot Left Joystick
   - Moving in the x-y plane
 - Pilot Right Joystick
   - Rotation around the robots axis
 - Pilot.a()
   - Reset Heading

### CODE REVIEW

File: Team 3749 2023/src/main/java/frc/robot/Xbox.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package frc.robot.utils;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/Xbox.java]: why is this outside of utils?

File: Team 3749 2023/src/main/java/frc/robot/Xbox.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
...
+package frc.robot;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/Xbox.java]: update this class to use the 2023 CommandXboxController
https://github.com/Team-Optix-3749/Team-3749-2023/blob/arm/Team%203749%202023/src/main/java/frc/robot/utils/Xbox.java

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingHellaSimple.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+/***
+ * @author Noah Simon
+ * 
+ *         Allows the robot to automatically "engage" on the charging station
+ */
+public class AutoBalancingHellaSimple extends CommandBase {
+    @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+
+    SwerveSubsystem swerveSubsystem;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingHellaSimple.java:36]: make your class variables private & final

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingHellaSimple.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        swerveSubsystem.stopModules();
+    }
+
+    // Returns true when the command should end
+    @Override
+    public boolean isFinished() {
+        return false;
+    }
+
+    /***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingHellaSimple.java:145]: move this to above the overridden methods

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingPID.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+/***
+ * @author Noah Simon
+ * 
+ *         Allows the robot to automatically "engage" on the charging station
+ */
+public class AutoBalancingPID extends CommandBase {
+    @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+
+    SwerveSubsystem swerveSubsystem;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingPID.java:36]: private and final

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingPID.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        swerveSubsystem.stopModules();
+    }
+
+    // Returns true when the command should end
+    @Override
+    public boolean isFinished() {
+        return false;
+    }
+
+    /***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingPID.java:150]: move this to above the overridden methods

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingHellaSimple.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.trajectory.Trajectory;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+/***
+ * @author Noah Simon
+ * 
+ *         Allows the robot to automatically "engage" on the charging station
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingHellaSimple.java:30]: Explain the diff between ur different autobalancing commands here. what does toby do? what does PID do? what does hellasimple do?

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingToby.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+/***
+ * @author Noah Simon
+ * 
+ *         Allows the robot to automatically "engage" on the charging station
+ */
+public class AutoBalancingToby extends CommandBase {
+    @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+
+    SwerveSubsystem swerveSubsystem;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingToby.java:35]: private and final

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingToby.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    /***
+     * 
+     * @param margin how close the values need to be to return true. Use a positive
+     *               number
+     * @param a      the first number
+     * @param b      the second number
+     * @return true if it is within the margin, false if not
+     */
+    private boolean withinMargin(double margin, double a, double b) {
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoBalancingToby.java:154]: move above

File: Team 3749 2023/src/main/java/frc/robot/commands/AutoCommands.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SwerveControllerCommand;
+import frc.robot.Constants;
+import frc.robot.subsystems.SwerveSubsystem;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+
+/***
+ * A class to return command sequences for auto
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/AutoCommands.java]: add link to [pp docs](https://github.com/mjansen4857/pathplanner/wiki/PathPlannerLib:-Java-Usage)

File: Team 3749 2023/src/main/java/frc/robot/commands/MoveDistance.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+/***
+ * @author Noah Simon
+ * 
+ *         Moves the robot a specific amount forwrad on a button press
+ */
+public class MoveDistance extends CommandBase {
+    @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+
+    SwerveSubsystem swerveSubsystem;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/MoveDistance.java:21]: private final

File: Team 3749 2023/src/main/java/frc/robot/RobotContainer.java [hunk 12]
```diff
@@ -4,25 +4,82 @@
...
   }
 
-  private void configureBindings() {
-    final Xbox pilot = new Xbox(0);
-    final Xbox operator = new Xbox(1);
+  private void configureDefaultCommands() {
+  }
+
+  private void configureButtonBindings() {
+    pilot.a().whileTrue(new InstantCommand(swerveSubsystem::zeroHeading));
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/RobotContainer.java]: update these bindings with the new Xbox class

File: Team 3749 2023/src/main/java/frc/robot/commands/SwerveJoystickCommand.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+import edu.wpi.first.math.filter.SlewRateLimiter;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Constants.DriveConstants;
+import frc.robot.Constants.OIConstants;
+import frc.robot.subsystems.SwerveSubsystem;
+
+public class SwerveJoystickCommand extends CommandBase {
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/SwerveJoystickCommand.java]: rename this to SwerveTeleopCommand as that is what we call [default commands within the Arm subsystem](https://github.com/Team-Optix-3749/Team-3749-2023/blob/arm/Team%203749%202023/src/main/java/frc/robot/commands/ArmTeleopCommand.java)

File: Team 3749 2023/src/main/java/frc/robot/commands/feedForwardTesting.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+import edu.wpi.first.math.trajectory.Trajectory;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+/***
+ * @author Noah Simon
+ * 
+ *         Moves the robot a specific amount forwrad on a button press
+ */
+public class feedForwardTesting extends CommandBase {
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/feedForwardTesting.java]: capitalize first initial

File: Team 3749 2023/src/main/java/frc/robot/commands/feedForwardTesting.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+
+    SwerveSubsystem swerveSubsystem;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/feedForwardTesting.java]: private and final

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Arm.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Arm.java:1]: delete this file

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
+
+package frc.robot.subsystems;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java:2]: delete this file

File: Team 3749 2023/src/main/java/frc/robot/subsystems/SwerveSubsystem.java [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+import edu.wpi.first.math.trajectory.Trajectory;
+import edu.wpi.first.wpilibj.SPI;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Constants.DriveConstants;
+import frc.robot.commands.AutoCommands;
+import frc.robot.utils.AprilTagGetters;
+
+public class SwerveSubsystem extends SubsystemBase {
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/SwerveSubsystem.java:34]: add author and class javadoc comment

File: Team 3749 2023/src/main/java/frc/robot/utils/AprilTagGetters.java [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+import edu.wpi.first.math.geometry.Translation3d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import frc.robot.Constants;
+import frc.robot.Constants.*;
+
+/**
+ * @author Advik Garg
+ * @author Akhil Singamneni
+ * 
+ *
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/utils/AprilTagGetters.java]: delete empty lines & add class description

File: Team 3749 2023/src/main/java/frc/robot/utils/Limelight.java [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+public class Limelight {
+    private static NetworkTableEntry tx;
+    private static NetworkTableEntry ta;
+    public Limelight(NetworkTableEntry tx, NetworkTableEntry ta) {
+        NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight"); // New network table instance, and reading in the data table from limelight.
+        this.tx = tx; // Horizontal offset from crosshair to the target
+        this.ta = ta; // Target area
+
+    }
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/utils/Limelight.java:15]: getters?????

File: Team 3749 2023/src/main/java/frc/robot/utils/LimelightPhotonGetters.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
+package frc.robot.utils;
+
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import frc.robot.Constants.VisionConstants;
+
+/**
+ * @author Advik Garg
+ * @author William Cheng
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/utils/LimelightPhotonGetters.java:10]: add class description

File: Team 3749 2023/vendordeps/REVLib.json [hunk 22]
```diff
@@ -1,7 +1,7 @@
 {
```
Comment (rohinsood) [inline Team 3749 2023/vendordeps/REVLib.json]: [update revlib](https://software-metadata.revrobotics.com/REVLib-2023.json)

File: Team 3749 2023/src/main/java/frc/robot/RobotContainer.java [hunk 23]
```diff
@@ -4,25 +4,82 @@
...
 
 public class RobotContainer {
+
+  // Controllers
+  private final Xbox pilot = new Xbox(OIConstants.kDriverControllerPort);
+  private final Xbox operator = new Xbox(1);
+
+  // private final POV pilotPOV = new POV(pilot);
+  // private final POV operatorPOV = new POV(operator);
+  private final Joystick driverJoytick = new Joystick(OIConstants.kDriverControllerPort);
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/RobotContainer.java]: with the new Xbox class this is unecessary

**Review by rohinsood — COMMENTED** [review]
Great code code code! Some formatting standardization but other than that theres no major changes that I can see

**Comment by raad1masum** [comment]
Auto balance and vision code are not working. However, I will merge this PR and delete the according files in the branch afterward. Next time this PR should be merged before moving on to new stuff.

**Review by raad1masum — APPROVED** [review]

**Review by rohinsood — APPROVED** [review]
Code code code!

**Pull request merged**